yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnAllocatorSuite.scala

```diff
@@ -242,4 +245,154 @@ class YarnAllocatorSuite extends SparkFunSuite with Matchers with BeforeAndAfter
     assert(pmemMsg.contains("2.1 MB of 2 GB physical memory used."))
   }
 
+  test("allocate locality preferred containers with enough resource and no matched existed " +
+    "containers") {
+    // 1. All the locations of current containers cannot satisfy the new requirements
+    // 2. Current requested container number can fully satisfy the pending tasks.
+
+    val handler = createAllocator(2)
+    handler.updateResourceRequests()
+    handler.handleAllocatedContainers(Array(
+      createContainer("host1"),
+      createContainer("host2")))
+
+    handler.requestTotalExecutorsWithPreferredLocalities(
+      5, 15, Map("host3" -> 15, "host4" -> 15, "host5" -> 10))
+    handler.updateResourceRequests()
+    val nodesOfRequests = handler.containerRequests.toArray.map { request =>
+      if (request.getNodes == null) {
+        null
+      } else {
+        request.getNodes.toArray(new Array[String](request.getNodes.size))
+      }
+    }
+
+    assert(nodesOfRequests === Array(
+      null, // container1 request
+      null, // container2 request
```

Comment:
Hi @squito , sorry for it, the first and second `null` is the node list of `ContainerRequest` here:

```
 val handler = createAllocator(2)
    handler.updateResourceRequests()
    handler.handleAllocatedContainers(Array(
      createContainer("host1"),
      createContainer("host2")))
```

Here at start, I created two `ContainerRequest` with preferred node list `null`, but actually allocated on "host1" and "host2", so the preferred node list of these two `ContainerRequest` are `null`, I'm not sure if I explained clearly, I will add more comments.